File: homework/tai_nguyen/lesson7/113_path_sum_ii.ts
/**
 * @link https://leetcode.com/problems/path-sum-ii/
 */

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function isLeaf(node: TreeNode) {
  return node.left === null && node.right === null;
}

function pathSum(root: TreeNode | null, targetSum: number): number[][] {
  if (root === null) return [];

  if (isLeaf(root)) {
    if (root.val === targetSum) {
      return [[root.val]];
    }

    return [];
  }

  const leftPaths = pathSum(root.left, targetSum - root.val);
  const rightPaths = pathSum(root.right, targetSum - root.val);

  return leftPaths.concat(rightPaths).map((path) => [root.val].concat(path));
}

File: homework/thinhbka/lesson5-graph/113._Path_Sum_II.py
# https://leetcode.com/problems/path-sum-ii/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
#Time O(n)
#Space O (n)
class Solution:
    def dfs(self,node,curr_arr,target):
        if node:
            if node.left is None and node.right is None:
                if node.val == target:
                    curr_arr.append(node.val)
                    self.ans.append(curr_arr.copy())
                    curr_arr.pop()
            else:
                curr_arr.append(node.val)
                new_target = target-node.val
                self.dfs(node.left,curr_arr,new_target)
                self.dfs(node.right,curr_arr,new_target)
                curr_arr.pop()
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> List[List[int]]:
        self.ans = []
        self.dfs(root,[],targetSum)
        return self.ans